### PR TITLE
add yolov3 for dygraph

### PR DESCRIPTION
**Pull request by tink2123**
**Add yolov3 model for dygraph**

* todo:
  1. upload pretrain model
  2. polish readme

### CODE REVIEW

File: dygraph/yolov3/README.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    sh ./weights/download.sh
+
+**注意：** Windows用户可通过`./weights/download.sh`中的链接直接下载和解压。
+
+通过设置`--pretrain` 加载预训练模型。同时在fine-tune时也采用该设置加载已训练模型。
+请在训练前确认预训练模型下载与加载正确，否则训练过程中损失可能会出现NAN。
+
+**开始训练：** 数据准备完毕后，可以通过如下的方式启动训练：
+
+    python train_dy.py \
```
Comment (heavengate) [inline dygraph/yolov3/README.md]: dy_graph 下的就直接用train.py吧

File: dygraph/yolov3/README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+```
+Iter 0, loss 4754.930664, time 104.66020
+Iter 0, loss 4872.500000, time 108.11873
+Iter 0, loss 6561.949707, time 98.08920
+Iter 0, loss 10159.562500, time 111.02826
+Iter 1, loss 7538.894043, time 8.89545
+Iter 1, loss 7853.537109, time 7.02231
+Iter 1, loss 10502.829346, time 19.90573
+Iter 1, loss 8121.074707, time 13.60313
```
Comment (heavengate) [inline dygraph/yolov3/README.md]: 训练速度这么慢么

File: dygraph/yolov3/README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    python eval_dy.py \
+        --dataset=coco2017 \
+        --weights=${path_to_weights} \
+        --class_num=${category_num}
+
+- 通过设置`export CUDA_VISIBLE_DEVICES=0`指定单卡GPU评估。
+
+
+若训练时指定`--syncbn=True`, 模型评估精度如下:
```
Comment (heavengate) [inline dygraph/yolov3/README.md]: 动态图没有syncbn？

File: dygraph/yolov3/edict.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserve.
```
Comment (heavengate) [inline dygraph/yolov3/edict.py]: 2019 -> 2020

File: dygraph/yolov3/train_dy.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+from paddle.fluid.dygraph.base import to_variable
+
+num_trainers = int(os.environ.get('PADDLE_TRAINERS_NUM', 1))
+
+
+def get_device_num():
+    # NOTE(zcd): for multi-processe training, each process use one GPU card.
+    if num_trainers > 1:
+        return 1
+    return fluid.core.get_cuda_device_count()
```
Comment (heavengate) [inline dygraph/yolov3/train_dy.py]: 动态图现在能在CPU和windows上跑么，能的话要考虑兼容性，最近yolov3里有几个关于CPU和windows的兼容性pr合入，可以同步一下

File: dygraph/yolov3/train_dy.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    if not os.path.exists(cfg.model_save_dir):
+        os.makedirs(cfg.model_save_dir)
+
+    gpu_id = int(os.environ.get('FLAGS_selected_gpus', 0))
+    place = fluid.CUDAPlace(fluid.dygraph.parallel.Env().dev_id) if cfg.use_data_parallel else fluid.CUDAPlace(0)
+
+    with fluid.dygraph.guard(place):
+        if args.use_data_parallel:
+            strategy = fluid.dygraph.parallel.prepare_context()
+        model = Yolov3(3, is_train=True)
```
Comment (heavengate) [inline dygraph/yolov3/train_dy.py]: Yolov3 -> YOLOv3

File: dygraph/yolov3/train_dy.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                learning_rate=lr,
+                warmup_steps=cfg.warm_up_iter,
+                start_lr=0.0,
+                end_lr=cfg.learning_rate)
+
+        optimizer = fluid.optimizer.Momentum(
+            learning_rate=lr,
+            regularization=fluid.regularizer.L2Decay(cfg.weight_decay),
+            momentum=cfg.momentum,
+            #parameter_list=model.parameters()
```
Comment (heavengate) [inline dygraph/yolov3/train_dy.py]: delete or uncomment

File: dygraph/yolov3/README.md (hunk 2 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: 第一个iter带了启动时间，会慢一点

File: dygraph/yolov3/README.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+### 模型评估
+
+模型评估是指对训练完毕的模型评估各类性能指标。本示例采用[COCO官方评估](http://cocodataset.org/#detections-eval), 用户可通过如下方式下载Paddle发布的YOLOv3[模型]()
+
+    sh ./weights/download.sh
+
+`eval_dy.py`是评估模块的主要执行程序，调用示例如下：
+
+    python eval_dy.py \
```
Comment (heavengate) [inline dygraph/yolov3/README.md]: 同上

**Review by heavengate — COMMENTED** [review]
infer代码也补充下吧

File: dygraph/yolov3/train_dy.py (hunk 5 above)
Comment (tink2123) [reply dygraph/yolov3/train_dy.py]: 可以在cpu上跑，windows没有验证过，这里我再确认一下

File: dygraph/yolov3/README.md (hunk 1 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: done

File: dygraph/yolov3/README.md (hunk 3 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: 动态图暂时没有加syncbn训练，readme已修改

File: dygraph/yolov3/train_dy.py (hunk 6 above)
Comment (tink2123) [reply dygraph/yolov3/train_dy.py]: done

File: dygraph/yolov3/train_dy.py (hunk 7 above)
Comment (tink2123) [reply dygraph/yolov3/train_dy.py]: 1.7版本的optimizer将需要指定parameter_list，暂时保留在这里，等合入时去掉注释

File: dygraph/yolov3/README.md (hunk 8 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: done

**Comment by tink2123** [comment]
> infer代码也补充下吧
done，thx！

**Comment by tink2123** [comment]
> infer代码也补充下吧

已添加～ 感谢review

File: dygraph/yolov3/README.md (hunk 2 above)
Comment (heavengate) [reply dygraph/yolov3/README.md]: Iter数没更新，另外最后一个iter是13s？

Comment (tink2123) [reply dygraph/yolov3/README.md]: 动态图打印log时，会将每张卡上的log都打印出来，这里用了4卡训练，所以每个iter会打印四次。最后一个iter是13s，后面iter的训练时间如下：
```
Iter 58251, loss 85.895201, time 1.90960
Iter 58251, loss 85.928059, time 2.03652
Iter 58251, loss 85.939094, time 2.91509
Iter 58251, loss 85.784909, time 2.94787
Iter 58252, loss 85.927157, time 2.11885
Iter 58252, loss 85.894701, time 2.51081
Iter 58252, loss 85.939130, time 4.03091
Iter 58252, loss 85.784688, time 3.34257
Iter 58253, loss 85.894178, time 2.63775
Iter 58253, loss 85.927063, time 3.69755
Iter 58253, loss 85.938522, time 2.64180
Iter 58253, loss 85.784001, time 3.09755
```

File: dygraph/yolov3/box_utils.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    inter_y1 = np.maximum(b1_y1, b2_y1)
+    inter_y2 = np.minimum(b1_y2, b2_y2)
+    inter_w = inter_x2 - inter_x1 + 1
+    inter_h = inter_y2 - inter_y1 + 1
+    inter_w[inter_w < 0] = 0
+    inter_h[inter_h < 0] = 0
+
+    inter_area = inter_w * inter_h
+    b1_area = (b1_x2 - b1_x1 + 1) * (b1_y2 - b1_y1 + 1)
+    b2_area = (b2_x2 - b2_x1 + 1) * (b2_y2 - b2_y1 + 1)
```
Comment (SunGaofeng) [inline dygraph/yolov3/box_utils.py]: +1 应该要去掉吧？

1、这个函数被调用的时候，box1和box2用的都是相对坐标，计算宽度的时候不用再加1.0了
2、对于xywh这种格式，即使使用的不是相对坐标，也不需要加1.0，
      按上面的计算过程：b1_x2 - b1_x1 + 1 = box1[:, 2] + 1，多算了一个像素点了

File: dygraph/yolov3/README.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+    git clone https://github.com/cocodataset/cocoapi.git
+    cd cocoapi/PythonAPI
+    # if cython is not installed
+    pip install Cython
+    # Install into global site-packages
+    make install
+    # Alternatively, if you do not have permissions or prefer
+    # not to install the COCO API into global site-packages
+    python2 setup.py install --user
```
Comment (SunGaofeng) [inline dygraph/yolov3/README.md]: 这里是只能用python2吗？如果不是的话，把python2换成python吧

File: dygraph/yolov3/README.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+**模型设置：**
+
+*  模型使用了基于COCO数据集生成的9个先验框：10x13，16x30，33x23，30x61，62x45，59x119，116x90，156x198，373x326
+*  YOLOv3模型中，若预测框不是该点最佳匹配框但是和任一ground truth框的重叠大于`ignore_thresh=0.7`，则忽略该预测框的目标性损失
+
+**训练策略：**
+
+*  采用momentum优化算法训练YOLOv3，momentum=0.9。
+*  学习率采用warmup算法，前4000轮学习率从0.0线性增加至0.001。在400000，450000轮时使用0.1,0.01乘子进行学习率衰减，最大训练500000轮。
```
Comment (SunGaofeng) [inline dygraph/yolov3/README.md]: 训练1轮，一般指的是训练1个epoch吧？这里是训练一个iter的话，是不是把“轮”换成“迭代”或者iter会准确点？

File: dygraph/yolov3/README.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+下图为模型训练结果：
+<p align="center">
+<img src="image/train_loss.png" height="400" width="550" hspace="10"/><br />
+Train Loss
+</p>
+
+### 模型评估
+
+模型评估是指对训练完毕的模型评估各类性能指标。本示例采用[COCO官方评估](http://cocodataset.org/#detections-eval), 用户可通过如下方式下载Paddle发布的YOLOv3[模型]()
```
Comment (SunGaofeng) [inline dygraph/yolov3/README.md]: 添加模型参数链接

File: dygraph/yolov3/README.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        --class_num=${category_num}
+
+- 通过设置`export CUDA_VISIBLE_DEVICES=0`指定单卡GPU评估。
+
+
+## 进阶使用
+
+### 背景介绍
+
+传统目标检测方法通过两阶段检测，第一阶段生成预选框，第二阶段对预选框进行分类得到类别，而YOLO将目标检测看做是对框位置和类别概率的一个单阶段回归问题，使得YOLO能达到近两倍的检测速度。而YOLOv3在YOLO的基础上引入的多尺度预测，使得YOLOv3网络对于小物体的检测精度大幅提高。
```
Comment (SunGaofeng) [inline dygraph/yolov3/README.md]: 第二阶段不仅仅是对预选框进行分类，也会对预选框位置坐标进行预测。

第一阶段生成预选框，第二阶段对预选框进行分类和位置坐标的调整

File: dygraph/yolov3/eval.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    with fluid.dygraph.guard(place):
+        model = YOLOv3(3,is_train=False)
+        # yapf: disable
+        if cfg.weights:
+            restore, _ = fluid.load_dygraph(cfg.weights)
+            model.set_dict(restore)
+            model.eval()
+
+        input_size = cfg.input_size
+        test_reader = reader.test(input_size, 1)
```
Comment (SunGaofeng) [inline dygraph/yolov3/eval.py:57]: 这地方batchsize=1，在预测时使用到了fluid.layers.multiclass_nms，其返回值是lod tensor，在动态图里面lod信息被丢失了。

如果batchsize>1的话，计算结果会不对，建议添加一行注释告诉用户只能使用batchsize=1

File: dygraph/yolov3/eval.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+            im_id = to_variable(im_id_data)
+
+            im_shape_data = np.array([x[2] for x in data]).astype('int32')
+            im_shape = to_variable(im_shape_data)
+
+            batch_outputs = model(img, None, None, None, im_id, im_shape)
+            nmsed_boxes = batch_outputs.numpy()
+            if nmsed_boxes.shape[1] != 6:
+                continue
+            for i in range(1):
```
Comment (SunGaofeng) [inline dygraph/yolov3/eval.py]: 可以把for I in range(1)删掉，调整下代码格式

File: dygraph/yolov3/box_utils.py (hunk 9 above)
Comment (tink2123) [reply dygraph/yolov3/box_utils.py]: done,thanks!

File: dygraph/yolov3/README.md (hunk 10 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: done, thanks!

File: dygraph/yolov3/README.md (hunk 11 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: done,Thanks!

File: dygraph/yolov3/README.md (hunk 12 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: 暂时没有上传训练好的模型，先删除了文字描述，待模型上传后更新readme

File: dygraph/yolov3/README.md (hunk 13 above)
Comment (tink2123) [reply dygraph/yolov3/README.md]: done,thanks!

File: dygraph/yolov3/eval.py (hunk 14 above)
Comment (tink2123) [reply dygraph/yolov3/eval.py:57]: done,thanks

File: dygraph/yolov3/eval.py (hunk 15 above)
Comment (tink2123) [reply dygraph/yolov3/eval.py]: done thanks!

**Review by heavengate — APPROVED** [review]
LGTM

**Comment by tink2123** [comment]
yolo_voc 在6w轮的eval结果：
image
